src/fp/list.nim

```diff
@@ -114,6 +114,24 @@ proc foldRightF*[T, U](xs: List[T], z: () -> U, f: (T, () -> U) -> U): U =
   if xs.isEmpty: z()
   else: f(xs.head, () => xs.tail.foldRightF(z, f))
 
+proc unfoldLeft*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the left from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = b ^^ result
+
+proc unfoldRight*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the right from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = result.append(b.asList)
```

Comment:
`b.asList` seems dangerous here, because `asList` is overloaded. For example, if `b` is an `Option`, this probably wouldn't compile. Use `Cons` or `b.point(List)` instead.